### PR TITLE
Fix a use-after-free and a shutdown crash

### PR DESCRIPTION
**Pull request by cataphract**
See the commit messages for more details.

### CODE REVIEW

**Comment by cataphract** [comment]
I forgot, this also fixes a leak; the threads created to destroy contexts were not detached, so they were expected to be joined.

File: ext/mini_racer_extension/mini_racer_extension.cc [hunk 1]
```diff
@@ -1235,22 +1253,17 @@ static void free_context(ContextInfo* context_info) {
...
     }
 
-    if (context_info->context && isolate_info && isolate_info->isolate) {
-        context_info->context = NULL;
-    }
-
-    if (isolate_info) {
-        context_info->isolate_info = NULL;
-    }
+    context_info->context = NULL;
```
Comment (SamSaffron) [inline ext/mini_racer_extension/mini_racer_extension.cc:1265]: just confirming we don't need the guards anymore?

**Comment by SamSaffron** [comment]
@cataphract  how would you feel about working on getting mini racer working with latest libv8 

We are struggling here: https://github.com/rubyjs/mini_racer/pull/153#issuecomment-623561107

File: ext/mini_racer_extension/mini_racer_extension.cc (hunk 1 above)
Comment (cataphract) [reply ext/mini_racer_extension/mini_racer_extension.cc:1265]: Yes, the problem is that calling `dispose`  on the context triggers the `IsolateInfo` to be deleted (through `free_context_raw`), but afterwards we attempt to access `context_info->isolate_info->isolate`.

**Comment by SamSaffron** [comment]
cool, lets try this out. Thanks!

**Comment by SamSaffron** [comment]
One big issue with this implementation is that it is fork unsafe. I can reproduce hangs on local with forked processes. 

This is particularity a big issue with my experimentation around single threaded platform: 

https://github.com/rubyjs/mini_racer/tree/single_threaded_platform

we are going to need to figure out a way of making this fork safe.

**Comment by cataphract** [comment]
@SamSaffron can you provide the code to reproduce it? Is it the problem that the mutex is in a bad state after the fork?

**Comment by SamSaffron** [comment]
Sure @cataphract give a shot in https://github.com/rubyjs/mini_racer/tree/single_threaded_platform

specifically test/test_forking.rb hangs for me. 

The unfortunate thing here though is that you are going to have to build a custom v8 for this to hang on exit https://github.com/rubyjs/mini_racer/issues/175#issuecomment-705898043

**Comment by cataphract** [comment]
@SamSaffron ok, I'll look at it this week

**Comment by SamSaffron** [comment]
thanks @cataphract !

**Comment by cataphract** [comment]
@SamSaffron I've added a PR: https://github.com/rubyjs/mini_racer/pull/180

**Pull request merged**